motionplan/planManager.go

```diff
@@ -0,0 +1,382 @@
+	}
+
+	// Build planner
+	var pathPlanner motionPlanner
+	if seed, ok := opt.extra["rseed"].(int); ok {
+		//nolint: gosec
+		pathPlanner, err = opt.PlannerConstructor(mp.frame, runtime.NumCPU()/2, rand.New(rand.NewSource(int64(seed))), mp.logger)
+	} else {
+		//nolint: gosec
+		pathPlanner, err = opt.PlannerConstructor(mp.frame, runtime.NumCPU()/2, rand.New(rand.NewSource(int64(mp.randseed.Int()))), mp.logger)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if rm == nil {
+		rm = initRRTMaps()
+	}
+
+	planctx, cancel := context.WithTimeout(ctx, time.Duration(opt.Timeout*float64(time.Second)))
+	defer cancel()
+
+	remainingSteps := [][]referenceframe.Input{}
+	if parPlan, ok := pathPlanner.(rrtParallelPlanner); ok {
+		// rrtParallelPlanner supports solution look-ahead for parallel waypoint solving
+		endpointPreview := make(chan node, 1)
+		solutionChan := make(chan *rrtPlanReturn, 1)
+		utils.PanicCapturingGo(func() {
+			parPlan.rrtBackgroundRunner(planctx, goal, seed, &rrtParallelPlannerShared{opt, rm, endpointPreview, solutionChan})
+		})
+		for {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+			select {
+			case nextSeed := <-endpointPreview:
+				// Got a solution preview, start solving the next motion in a new thread.
+				if iter+1 < len(goals) {
+					// In this case, we create the next step (and thus the remaining steps) and the
+					// step from our iteration hangs out in the channel buffer until we're done with it.
+					remainingSteps, err = mp.planMotion(ctx, goals, nextSeed.Q(), opts, nil, iter+1)
+					if err != nil {
+						return nil, err
+					}
+				}
+				for {
+					// Get the step from this runner invocation, and return everything in order.
+					select {
+					case <-ctx.Done():
+						return nil, ctx.Err()
+					default:
+					}
+
+					select {
+					case finalSteps := <-solutionChan:
+						if finalSteps.err() != nil {
+							return nil, finalSteps.err()
+						}
+						results := append(finalSteps.toInputs(), remainingSteps...)
+						return results, nil
+					default:
+					}
+				}
+			case finalSteps := <-solutionChan:
+				// We didn't get a solution preview (possible error), so we get and process the full step set and error.
+				if finalSteps.err() != nil {
+					if opt.Fallback != nil {
+						remainingSteps, err = mp.planMotion(
+							ctx,
+							goals,
+							seed,
+							append([]*plannerOptions{opt.Fallback}, opts[1:]...),
+							finalSteps.rm,
+							iter,
+						)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						return nil, finalSteps.err()
+					}
+				} else {
+					if iter+1 < len(goals) {
+						// in this case, we create the next step (and thus the remaining steps) and the
+						// step from our iteration hangs out in the channel buffer until we're done with it
+						remainingSteps, err = mp.planMotion(
+							ctx,
+							goals,
+							finalSteps.steps[len(finalSteps.steps)-1].Q(),
+							opts,
+							nil,
```

Comment:
General question: What's the thought behind not passing in `finalSteps.rm` here? Is there a particular reason you want a fresh set of RRT maps on the next call to `planMotion` in this case?

Reply:
This is the case where we found a solution. So the current set of maps are not valid and we want to restart.

We could remove just the seed/solutions from the maps, then verify branches that can be re-added. But IMO that's probably nearly as expensive as just re-doing.

Reply:
Sounds good